### ISSUE TITLE
Release 2.12.2

### ISSUE THREAD

**Issue by adriaanm** [scala/scala-dev#331]
we'll document the steps taken in releasing 2.12.2 here (taken from our currently internal release checklist)

**Comment by SethTisue** [comment]
there is a checklist now at https://github.com/scala/scala-dev/issues/332 that I can copy here when the time comes.

some additional notes:

* ~~2.11.9~~ 2.11.10 is a prerequisite. (in general, but also specifically: case class apply/unapply, SCP-009 direct-dependency work)
* draft release notes ahead of time, not at the end of the process
* triage open JIRA tickets listed as 2.12.2
* triage open scala-dev tickets listed as 2.12.2
* update LICENSE in root dir
* check in with @dwijnand about trailing commas stuff (#299); we want to make sure that all rolls out smoothly
* before releasing, tell the community which nightly we intend to rebrand as 2.12.2, and invite them to test it

**Comment by SethTisue** [comment]
status update: 2.11.9 is on Maven Central but not announced (while Adriaan looks into a possible regression). we're still expecting/hoping to release 2.12.2 next week after merging a few additional PRs. the shape of the release can be seen from the [2.12.2 milestone](https://github.com/scala/scala/milestone/59), especially [the PRs with the release-notes tag](https://github.com/scala/scala/pulls?q=is%3Apr+label%3Arelease-notes+milestone%3A2.12.2)

**Comment by xuwei-k** [comment]
There is a same issue as Scala 2.11.9 (broken forward binary compatibility)

- https://github.com/xuwei-k/Scala-2.11.9-FilteredTraversableInternal/commit/ad1260360053b9335363ad861e98c1256bdc5314
- https://github.com/scala/scala/commit/eb5c51383a63c5c3420e53ef021607ff5fd20296

**Comment by adriaanm** [comment]
Thank you @xuwei-k. Glad you caught this before 2.12.2! I will revert the incompatible change.

**Comment by SethTisue** [comment]
most 2.12.2 JIRA tickets have now been assigned to "Backlog" or 2.12.3. four open tickets left at https://issues.scala-lang.org/issues/?jql=project%20%3D%20SI%20AND%20status%20%3D%20Open%20AND%20fixVersion%20%3D%20%22Scala%202.12.2%22%20ORDER%20BY%20status%20ASC , the relevant PRs are https://github.com/scala/scala/pull/5762 and https://github.com/scala/scala/pull/5402

**Comment by lrytz** [comment]
we should probably also merge 2.11.x to 2.12.x once 2.11.10 is tagged

**Comment by SethTisue** [comment]
@adriaanm I'm hoping you could do the 2.11->2.12 merge, since you're already spun up on what the changes are

**Comment by adriaanm** [comment]
Merge PR: https://github.com/scala/scala/pull/5837

**Comment by adriaanm** [comment]
- [X] I'll double check the bytecode. (for the record, scala/scala#5673, scala/scala#5663, scala/scala#5579 added bytecode-public methods)
  - [x] could someone else (@szeiger?) check the PRs I listed above and make sure we're still cool shipping them, knowing what we learned from 2.11.9?

Here's the full bytecode diff of the library from 2.12.1 to 2.12.2: https://gist.github.com/adriaanm/dccef51d7ef9aa2bb1c0688945d4c8cd

**Comment by SethTisue** [comment]
https://github.com/scala/scala/pull/5402 is merged. I pushed some community build tweaks and started a trial release-main job. current release candidate is scala/scala@15e2759f49a3ef0f71290f4bbd9839cbf2346b0e

**Comment by lrytz** [comment]
`pcplod` failed in a strange way (https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/1414/consoleFull)

```
[pcplod:error] scala.MatchError: fef15bff+20170411-0503-SNAPSHOT (of class java.lang.String)
[pcplod:error] 	at com.typesafe.dbuild.project.build.LocalBuildRunner$.prepareDepsArtifacts(LocalBuildRunner.scala:110)
```

maybe the switch to a new branch in https://github.com/scala/community-builds/commit/7032ae82a21c6055f479010d08cc0dca4efe2b08 caused the version number to change?

**Comment by dwijnand** [comment]
No the problem is the fork didn't retrain the tags which the version depends upon (via my sbt-dynver plugin).

**Comment by lrytz** [comment]
ach so! thanks :)

**Comment by xeno-by** [comment]
@SethTisue Has the trial build been published somewhere?

**Comment by lrytz** [comment]
@xeno-by `2.12.2-bin-15e2759` on https://scala-ci.typesafe.com/artifactory/scala-integration

**Comment by lrytz** [comment]
@SethTisue I pushed the tags from https://github.com/ensime/pcplod to https://github.com/scalacommunitybuild/pcplod and re-started the build at https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/1415/console

**Comment by SethTisue** [comment]
@dwijnand glad you figured that out because I had been baffled

@lrytz the new build failed in the same way. perhaps it wasn't a truly fresh build? I pushed an empty commit to the pcplod fork, trying again at https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/1416/consoleFull

UPDATE: yup, green now

**Comment by adriaanm** [comment]
I need to fix my breaking scalameta in https://github.com/scala/scala/pull/5816. PR forthcoming asap

**Comment by xeno-by** [comment]
@adriaanm Please let me know when you're done. I'll see whether your changes are compatible with paradise.

**Comment by adriaanm** [comment]
@xeno-by The revert in scala/scala#5844 hopefully makes 2.12.2 fully compatible with 2.12.1 for scala.meta. Let me know if I missed something!

**Comment by SethTisue** [comment]
I hit "merge" on 5844 and kicked off a new release-main job at https://scala-ci.typesafe.com/job/scala-2.12.x-release-main/756/consoleFull

**Comment by adriaanm** [comment]
The breakage discovered by akkasse  is tracked as https://github.com/scala/bug/issues/10261

**Comment by szeiger** [comment]
I took another look at https://github.com/scala/scala/pull/5673, https://github.com/scala/scala/pull/5663 and https://github.com/scala/scala/pull/5579. They look OK to me from a binary compatibility PoV.

**Comment by adriaanm** [comment]
🎈 Scala 2.12.2 is now *staged* at https://oss.sonatype.org/content/repositories/staging/. We intend to tag & promote to Maven Central next Tuesday, unless we hear of any regressions. 🙏

**Comment by sjrd** [comment]
FYI: I will be on vacation in New York from tomorrow until Sunday April 23. I should be able to release Scala.js for 2.12.2 at some point, but I might not be as reactive as usual.

**Comment by adriaanm** [comment]
Thanks for the heads up. Have a great time!

**Comment by adriaanm** [comment]
https://scala-ci.typesafe.com/job/scala-2.12.x-release-main/760/
  - https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-bootstrap/924/: 
    - scala/scala@21d12e9f5ec1ffe023f509848911476c1552d06f
    - Closed sonatype staging repos: https://oss.sonatype.org/content/repositories/orgscala-lang-1400.
  - https://scala-ci.typesafe.com/job/scala-2.12.x-release-package-unix/571/: scala/scala-dist@e49f4b3a8b9d83d21f7c50e283f5426808f21231

**Comment by adriaanm** [comment]
I created a v2.12.2 branch to address scala/bug#10263

**Comment by adriaanm** [comment]
Release-notes PRs: https://github.com/scala/scala/pulls?utf8=%E2%9C%93&q=is%3Apr%20label%3Arelease-notes%20is%3Amerged%20milestone%3A2.12.2

**Comment by xeno-by** [comment]
Macro paradise 2.1.0 for Scala 2.12.2 is now available at: https://oss.sonatype.org/content/repositories/orgscalamacros-1073/. Will release that staging repo once Scala 2.12.2 is released.

**Comment by xuwei-k** [comment]
https://github.com/scala/bug/issues/10268

**Comment by adriaanm** [comment]
Thanks for the report, @xuwei-k. As discussed in the ticket, the change was intentional.

**Comment by adriaanm** [comment]
🎉  Tags pushed and staging repo promoted!

**Comment by adriaanm** [comment]
Release announcement: https://github.com/scala/scala-lang/pull/636
Release notes: https://github.com/scala/scala/releases/tag/v2.12.2

**Comment by xeno-by** [comment]
Macro paradise 2.1.0 for Scala 2.12.2 has been released as well: https://oss.sonatype.org/content/repositories/releases/org/scalamacros/paradise_2.12.2/. Scala.meta / scala.meta paradise releases are pending on a new release of Scala.js /cc @sjrd

**Comment by SethTisue** [comment]
2.12.2 is on Maven Central and fully announced. version bump PR to follow

**Comment by sjrd** [comment]
Scala.js 0.6.15 released for Scala 2.12.2.

**Comment by xeno-by** [comment]
Meta & meta paradise 1.7.0 released for 2.12.2.
